### PR TITLE
nixos: configurable system-wide malloc

### PR DESCRIPTION
**Pull request by joachifm**
A more general alternative to https://github.com/NixOS/nixpkgs/pull/59955 that allows choosing among a predefined set of providers.  The implementation is pretty bad but does "work".  Ideally this would be improved upon later.

If this variant is deemed too much for the niche functionality it provides, I'm happy to go back to a more limited version under hardened. (From my pov, only the hardened allocator is really of interest).

### CODE REVIEW

**Comment by joachifm** [comment]
@GrahamcOfBorg test hardened

File: nixos/modules/config/malloc.nix [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  options = {
+    environment.memoryAllocator.provider = mkOption {
+      type = types.enum ([ "libc" ] ++ attrNames providers);
+      default = "libc";
+      description = ''
+        The system-wide memory allocator.
+        </para>
+
+        <para>
+        Briefly, the system-wide memory allcoator providers are:
```
Comment (ajs124) [inline nixos/modules/config/malloc.nix]: Typo: "allcoator"

File: nixos/modules/config/malloc.nix [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  # needlessly bloating the system closure
+  mallocLib = pkgs.runCommand "malloc-provider-${cfg.provider}"
+    rec {
+      preferLocalBuild = true;
+      allowSubstitutes = false;
+      origLibPath = providerConf.libPath;
+      libName = baseNameOf origLibPath;
+    }
+    ''
+    mkdir -p $out/lib
```
Comment (Lassulus) [inline nixos/modules/config/malloc.nix]: please indent

**Comment by eadwu** [comment]
`graphene-hardened` more or less just breaks `nix` for me, making it error out with `out of memory`.

**Comment by delroth** [comment]
+1, I just ended up tracking nix OOMs to this PR as well.

**Comment by delroth** [comment]
(If you're looking for another potential hardened allocator, Scudo has seen some good amounts of battle-testing. I suspect it might have less side effects.)

**Comment by joachifm** [comment]
See also https://github.com/NixOS/nixpkgs/issues/61489

**Pull request merged**